### PR TITLE
Add VFS util functions

### PR DESCRIPTION
**Pull request by dulanjalidilmi**
## Purpose
The purpose of this PR is to add VFS Util functions

## Approach
Add following util functions
* Function to upload a file to a FTP Server
* Function to download a file from a FTP Server
* Function to list the set of directories on a FTP server
* Function to list files of a directory on a FTP server
* Function to list directories & files from a directory on a FTP server
* Function to delete a file from a FTP server

## Test environment
This is tested locally in following environment.
JDK - 1.8
OS - macOS High Sierra

### CODE REVIEW

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
+package org.wso2.carbon.esb.scenario.test.common.utils;
```
Comment (milindaperera) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: Let's add license header

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.commons.net.ftp.FTP;
+import org.apache.commons.net.ftp.FTPClient;
+import org.apache.commons.net.ftp.FTPFile;
+
+public class VFSUtils {
```
Comment (milindaperera) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: Class level comment

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+     *
+     * @param hostname host name or IP address of the FTP server
+     * @param userName username of a FTP account on the FTP server to connect to
+     * @param password password corresponds to the username
+     * @param sourceFilePath source file path
+     * @param targetFilePath target path of the remote file
+     * @throws IOException if connecting to FTP server fails or file upload fails or logging out from ftp client fails
+     */
+    public static void ftpFileUpload(String hostname, String userName, String password, String sourceFilePath,
+                                     String targetFilePath) throws IOException {
```
Comment (milindaperera) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: Is it possible to provide port as well?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        FTPClient ftpClient = connectToFtp(hostname, userName, password);
+
+        try {
+            ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
+            File localFile = new File(sourceFilePath);
+
+            InputStream inputStream = new FileInputStream(localFile);
+            log.info("Start uploading file");
```
Comment (milindaperera) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: Shall we add the sourceFilePath for this log

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+import org.apache.commons.net.ftp.FTP;
+import org.apache.commons.net.ftp.FTPClient;
+import org.apache.commons.net.ftp.FTPFile;
+
+public class VFSUtils {
+    private static final Log log = LogFactory.getLog(VFSUtils.class);
+    private static int port = 21;
+
+    /**
+     * Function to upload a file to a FTP Server
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: to *an ftp server

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+     */
+    public static void ftpFileUpload(String hostname, String userName, String password, String sourceFilePath,
+                                     String targetFilePath) throws IOException {
+        FTPClient ftpClient = connectToFtp(hostname, userName, password);
+
+        try {
+            ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
+            File localFile = new File(sourceFilePath);
+
+            InputStream inputStream = new FileInputStream(localFile);
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: We need to close these streams if an error occurs as well. Hence, shall we create these in a try with resources block? Let's do it to all resources that implement the interface 'Closeable"

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java (hunk 3 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: Fixed. Added VFSClient class.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java (hunk 4 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: fixed

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java (hunk 6 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/utils/VFSUtils.java]: fixed.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * VFS client for vfs operations
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Doc comments need to end with a period.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    private String hostname;
+    private String userName;
+    private String password;
+    private int port;
+    private FTPClient ftpClient;
+
+    private static final Log log = LogFactory.getLog(VFSClient.class);
+
+    /**
+     * Constructor to VFSClient
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Doc comments need to end with a period. Shall we correct in all places?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+        try (InputStream inputStream = new FileInputStream(localFile)) {
+            log.info("Start uploading file " + sourceFilePath + " to " + targetFilePath);
+
+            try {
+                if (ftpClient.storeFile(targetFilePath, inputStream)) {
+                    log.info("File has been uploaded to " + targetFilePath + " successfully.");
+                }
+            } catch (IOException ex) {
+                throw new IOException("IO Exception occurred while uploading file ", ex);
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Do we really need to catch the exception and create a new exception? The method call will anyway be there in the stacktrace.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+     * Function to download a file from an FTP Server
+     *
+     * @param sourceFilePath source file path
+     * @param targetFilePath target path
+     * @throws IOException if file download fails
+     */
+    public void ftpFileDownload(String sourceFilePath, File targetFilePath) throws IOException{
+        ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
+
+        try (OutputStream outputStream1 = new BufferedOutputStream(new FileOutputStream(targetFilePath))) {
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Any reason to name this outputStream**1**?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            log.info("Start downloading file " + sourceFilePath + " to " + targetFilePath);
+
+            try {
+                if (ftpClient.retrieveFile(sourceFilePath, outputStream1)) {
+                    log.info("File has been uploaded to " + targetFilePath + " successfully.");
+                }
+            } catch (IOException ex) {
+                throw new IOException("IO Exception occurred while downloading file ", ex);
+            }
+        } catch (FileNotFoundException ex) {
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Let's just add the exception to the method signature. I suppose we do not gain anything by catching it and creating a new exception. Shall we correct in all places?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * Function to list files of a directory on an FTP server
+     *
+     * @param filePath file path to list files
+     * @return list of files
+     * @throws IOException if listing set of files fails
+     */
+    public List<FTPFile> listFiles(String filePath) throws IOException {
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: We return an **array** of FTPFile in listDirectories. Why do we return a **List** in this? Shal we make the 2 methods consistent?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+    /**
+     * Function to delete a file from an FTP server
+     *
+     * @param filePath  file path to delete
+     * @throws IOException if deleting the file fails
+     */
+    public void ftpDeleteFile(String filePath)
+            throws IOException {
+        try {
+            boolean deleted = ftpClient.deleteFile(filePath);
```
Comment (sdkottegoda) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Would it be a better idea to return the boolean indicating the status?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 8 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: fixed.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 7 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: fixed.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 11 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: In this function IO Exception would throw when downloading the file fails or unable to find the file. So isn't it better to throw the exception with a descriptive message?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 13 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: fixed

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 10 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: fixed

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 9 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: IO Exception would be thrown when downloading the file fails or unable to find the file. So isn't it better to throw the exception with a descriptive message?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+        try (OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(targetFilePath))) {
+            log.info("Start downloading file " + sourceFilePath + " to " + targetFilePath);
+
+            try {
+                if (ftpClient.retrieveFile(sourceFilePath, outputStream)) {
+                    log.info("File has been uploaded to " + targetFilePath + " successfully.");
+                }
+            } catch (IOException ex) {
+                throw new IOException("Error occurred while downloading file from " + sourceFilePath + " to " +
```
Comment (abeykoon) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Shall we break the line like below (+ should be at the begining of the line)

"Error occurred while downloading file from " + sourceFilePath 
 + " to + ...

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+        try (InputStream inputStream = new FileInputStream(localFile)) {
+            log.info("Start uploading file " + sourceFilePath + " to " + targetFilePath);
+
+            try {
+                if (ftpClient.storeFile(targetFilePath, inputStream)) {
+                    log.info("File has been uploaded to " + targetFilePath + " successfully.");
+                }
+            } catch (IOException ex) {
+                throw new IOException("Error occurred while uploading file from " + sourceFilePath + " to " +
```
Comment (abeykoon) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Shall we break the line like below (+ should be at the begining of the line)

```
"Error occurred while uploading file from " + sourceFilePath 
      + " to " +
```

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 14 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Fixed by PR [https://github.com/wso2/product-ei/pull/3466](https://github.com/wso2/product-ei/pull/3466)

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java (hunk 15 above)
Comment (dulanjalidilmi) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/carbon/esb/scenario/test/common/vfs/VFSClient.java]: Fixed by PR [https://github.com/wso2/product-ei/pull/3466](https://github.com/wso2/product-ei/pull/3466)

**Pull request merged**